### PR TITLE
[air] Store unflattened metrics in _TrackedCheckpoint

### LINKED ISSUE #35187: [AIR] Correctly unpack metrics with "/" in their names
### What happened + What you expected to happen

In #35000 , we unflattened the metrics for `trial.last_result`. However, if the metric name itself has a "/" character, it will be unflatten into a dictionary instead, which should not happen. One release test [failed](https://console.anyscale-staging.com/o/anyscale-internal/jobs/prodjob_4k4fdm16cci36ks4n4sna2hqjg) because of this change.

e.g.
```
metric = {"ptl/val_accuracy": 0.9}
# after dict unflatten
metric = {"ptl": {"val_accuracy": 0.9}}
```

One fact: The users should be able to use metrics with '/'. For example, it's meaningful for wandb because the metrics with the same prefix will be grouped together.

#### Potential Solution:

Instead of flatten into a string delimited by '/', we flatten into a tuple. 
```
Current way:
{"a": {"b": {"c": 1}}} -> {"a/b/c": 1}

After change:
{"a": {"b": {"c": 1}}} -> {("a", "b", "c")": 1}
```

### Versions / Dependencies

on master branch

### Reproduction script

`/Users/yunxuanx/Code/ray/release/lightning_tests/workloads/test_tuner.py`

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

We currently store metrics associated with a tracked AIR checkpoint in flattened format. This was done so we can specify nested keys in e.g. `checkpoint_score_attr`, which determines which checkpoints to keep.

However, we also rely on these stored metrics to restore the `trial.last_result` property when restoring from a checkpoint. Since #35000, we unflatten the dict there to resolve issues where we relied on the original structure in `trial.last_result`.

This again leads to issues with metrics in e.g. pytorch lightning, where some metrics are reported with a `/` (our default flatten separator). Or in a nutshell, `dict != unflatten(flatten(dict))` when the separator can be contained in the dict keys.

Some users may prefer to report with `/`, since it shows up nicely as nested metrics in wandb.

The proper resolution here is to not flatten the metrics at all. Instead, we should use `unflattened_lookup` to lookup the `checkpoint_score_attr` from the original metrics dict. 

This PR updates the `_TrackedCheckpoint` to retain the unflattened metrics.

## Related issue number

Closes #35187

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/_internal/checkpoint_manager.py [hunk 1]
```diff
@@ -382,15 +382,21 @@ def _get_checkpoint_score(
...
-                f"Result dict has no key: {checkpoint_score_attribute}. "
-                f"checkpoint_score_attr must be set to a key in the "
-                f"result dict. Valid keys are: {list(checkpoint.metrics.keys())}"
-            )
-            checkpoint_result = float("-inf")
+        if checkpoint_score_attribute:
+            try:
+                checkpoint_result = unflattened_lookup(
+                    checkpoint_score_attribute, checkpoint.metrics
+                )
```
Comment (justinvyu) [inline python/ray/air/_internal/checkpoint_manager.py:389]: I don't think this will work for using a metric containing `/`.

Unflattened dict:

```
metrics = {"a": {"b/c": 1.0}}
unflattened_lookup(metrics, "a/b/c") -> metrics["a"]["b"]["c"] not found
```

Comment (justinvyu) [reply python/ray/air/_internal/checkpoint_manager.py:389]: We can just flatten the metrics at this point, then access using the `checkpoint_score_attribute` (and don't keep the flattened copy around).

**Review by justinvyu — CHANGES_REQUESTED** [review]
Suggested one change!

I think that we should switch the API (not in this PR) to have the user pass in a tuple instead, so that this path concatenation is not as ambiguous:

```
checkpoint_config=air.CheckpointConfig(checkpoint_score_attribute=("a", "b/c"))
```

Same for `TuneConfig(metric)`.

File: python/ray/air/tests/test_checkpoint_manager.py [hunk 2]
```diff
@@ -97,6 +97,40 @@ def test_keep_best_checkpoints():
     ] == [1, 0]
 
 
+def test_nested_get_checkpoint_score():
```
Comment (justinvyu) [inline python/ray/air/tests/test_checkpoint_manager.py:100]: Can we also add the nested metric with `/` as a test case?

File: python/ray/air/_internal/checkpoint_manager.py (hunk 1 above)
Comment (krfricke) [reply python/ray/air/_internal/checkpoint_manager.py]: Yes, good idea!

File: python/ray/air/tests/test_checkpoint_manager.py (hunk 2 above)
Comment (krfricke) [reply python/ray/air/tests/test_checkpoint_manager.py]: Added!

**Comment by krfricke** [comment]
Thanks @justinvyu, great suggestions, added more test cases and better lookup!

File: python/ray/air/_internal/checkpoint_manager.py [hunk 3]
```diff
@@ -382,15 +382,20 @@ def _get_checkpoint_score(
...
+                valid_keys = list(flat_metrics.keys())
+                logger.error(
+                    f"Result dict has no key: {checkpoint_score_attribute}. "
+                    f"checkpoint_score_attr must be set to a key in the "
+                    f"result dict. Valid keys are: {valid_keys}"
+                )
+                checkpoint_result = float("-inf")
         else:
-            checkpoint_result = checkpoint.metrics[checkpoint_score_attribute]
+            checkpoint_result = float("-inf")
```
Comment (justinvyu) [inline python/ray/air/_internal/checkpoint_manager.py:398]: Should the default result should be the training iteration? We document that the earliest checkpoints get deleted if no `checkpoint_score_attribute` is given, but I think that just happens implicitly through the FIFO heap push/pop order -- maybe we should just always order the heap?

**Review by justinvyu — APPROVED** [review]
Thanks, one last question.

Can we also add a small docstring of the return type of `cpm._get_checkpoint_score`? A bit hard to parse the test assertion at the moment.

File: python/ray/air/_internal/checkpoint_manager.py (hunk 3 above)
Comment (krfricke) [reply python/ray/air/_internal/checkpoint_manager.py:409]: I've added a docstring that may shed some light on this. We actually explicitly order by checkpoint ID in the last tuple entry. The logic is as follows:

`_get_checkpoint_score` returns a tuple `(is_not_na, metric, id)`

Higher values are preferred/kept.

`sort()` first sorts by the first tuple index, then the second, then the third.

The first index ensures that NaN metrics are discarded first (rated worst).

The second index then sorts by actual metric.

If the metrics are the same (e.g. when no attr is defined --> we get -inf), then we sort by checkpoint ID and keep the later checkpoints.

Comment (justinvyu) [reply python/ray/air/_internal/checkpoint_manager.py:409]: Got it, that clears it up a lot!

**Comment by ArturNiederfahrenhorst** [comment]
CC @zhe-thoughts for approval, please 🙂

**Comment by ArturNiederfahrenhorst** [comment]
@krfricke Please make sure to also create the cherry pick PR yourself after this is merged!

**Comment by zhe-thoughts** [comment]
Yup @krfricke please merge this, create a cherry pick PR and assign to me and @ArturNiederfahrenhorst
